sherpa/ui/utils.py

```diff
@@ -319,103 +319,93 @@ def clean(self):
-
-        self._compsrcplot = sherpa.plot.ComponentSourcePlot()
-        self._compsrchistplot = sherpa.plot.ComponentSourceHistogramPlot()
 
         # self._comptmplmdlplot = sherpa.plot.ComponentTemplateModelPlot()
         self._comptmplsrcplot = sherpa.plot.ComponentTemplateSourcePlot()
 
-        self._sourceplot = sherpa.plot.SourcePlot()
-        self._sourcehistplot = sherpa.plot.SourceHistogramPlot()
-        self._fitplot = sherpa.plot.FitPlot()
-        self._residplot = sherpa.plot.ResidPlot()
-        self._delchiplot = sherpa.plot.DelchiPlot()
-        self._chisqrplot = sherpa.plot.ChisqrPlot()
-        self._ratioplot = sherpa.plot.RatioPlot()
-        self._psfplot = sherpa.plot.PSFPlot()
-        self._kernelplot = sherpa.plot.PSFKernelPlot()
         self._lrplot = sherpa.plot.LRHistogram()
         self._pdfplot = sherpa.plot.PDFPlot()
         self._cdfplot = sherpa.plot.CDFPlot()
         self._traceplot = sherpa.plot.TracePlot()
         self._scatterplot = sherpa.plot.ScatterPlot()
 
-        self._datacontour = sherpa.plot.DataContour()
-        self._modelcontour = sherpa.plot.ModelContour()
-        self._sourcecontour = sherpa.plot.SourceContour()
-        self._fitcontour = sherpa.plot.FitContour()
-        self._residcontour = sherpa.plot.ResidContour()
-        self._ratiocontour = sherpa.plot.RatioContour()
-        self._psfcontour = sherpa.plot.PSFContour()
-        self._kernelcontour = sherpa.plot.PSFKernelContour()
-
         self._intproj = sherpa.plot.IntervalProjection()
         self._intunc = sherpa.plot.IntervalUncertainty()
         self._regproj = sherpa.plot.RegionProjection()
         self._regunc = sherpa.plot.RegionUncertainty()
 
+        # The keys of _plot_types are used to define:
+        # a) the mapping from get_<key>_plot to the plot objects
+        #    (that is, there must be a matching get_<key>_plot method)
+        # b) valid arguments for the plot() call
+        # c) the arguments that set_xlog/... accept
+        # d) a set of names that can not be used as a dataset identifier
+        #    (because of point b); see also _contour_types
+        #
+        # Note that not all plot_xxx commands use this structure.
+        #
+        # Unlike the contour case, we have different plot classes to
+        # handle different types of plot. There are either plot types
+        # with only one plot class, such as "fit", or those that
+        # support a version that depends on the data type, such as
+        # "data". A list is used in both cases, with the first element
+        # being the "generic" case and, for those that support it, the
+        # second being the Data1DInt plot class.
+        #
+        # There is an argument to be made to have the "singleton"
+        # plots, such as "fit", storing the object directly, and not
+        # as a single-element list, but it was felt that having a
+        # consistent access pattern was cleaner.
```

Comment:
At least, I see the benefit of keeping a uniform representation here (always use a list), but this can be discussed.

Earlier attempts at fixing this actually created a class to use here (e.g. create an instance for the value rather than a list, and this class then had support for the "default" value and adding extra elements),  but I never found it satisfactory.

Reply:
This is not a full review, just some ideas based on what I'm seeing now; I want to make comments before it's all done an too late to change. So, please don't waste your time answering my suggestions if you already tried that, or think it's stupid and won't work. That said, here are some thoughts:

This entire approach was (and still is, although it's better now) pretty convoluted and requires careful maintenance. When new plot types are added, the `_plot_types` need to be updated, there is the `_plot_types` and `_contour_types` and later possible `_image_types` and then a few others that don't really fall in this scheme.

I am wondering if more of this information should live in the Plot classes themselves, e.g.:

```python
class RatioPlot(...):
    # would accept all Data1DInt and DataPA, but not generic Data1D
    # I know that that's not what RatioPlot does, but giving and example
    _accepted_data_types = (Data1DInt, DataPHA)  
    _plot_area = 'data'  # one of: data, contour, image
   _ plot_name = 'Ratio'

    ... real code here ...
```
(Obviously, all names are open to bike shedding.)

That way, information would not have to be maintained separately here, it would be intrinsic to the plot class. I think that's a better place to keep that information.

`_get_plot_type` would then search through all known plot classes

```python
def _get_plotorcountour_type(plotarea, plotname, data):
    for p in plot_type_list:
        if ((p._plot_area == plotarea) and (p._plot_name == plotname) and 
           any([insinstance(data, allowed_class) for allows_class in p._accepted_data_dypes])):
           return p
    finally:
        raise Error
```
The implementation above could be refined to deal with cases were there is more than one match, e.g. one could select the "most specific" of all plot classes (if one of them accepts all `Data1D` and another only `DataPHA`, then one may wnat to select the latter, while my implementation would pick whatever is first in the list).

How do we get the `plot_type_list`? Either that's hand-curated, but a better approach would be to use a metaclass that puts stuff in a registry when it's imported like I do for the plotting backends in #1382. That's just a few lines of code in the `Plot` and `Histogram` base classes (could be a new class in the tree `PLotHistroImageBase` from which `PLot` and `Histrogram` inherit).

Reply:
I just occurs to me that in practice
```python
    _accepted_data_types = (Data1DInt, DataPHA)  
    _plot_area = 'data'  # one of: data, contour, image, other
```
would not be set for every plot type, but at one point in the inheritance tree and then apply to all derived classes with no further code change. For example `Plot` would get the `    _plot_area = 'data' ` and all derived classes would have that property and `Contour` would get `_plot_area = 'contour'` and we only need to pay attention to special cases with multiple inheritance. 

I've overlooked something that needs to be special cased in the current scheme, then maybe it also has to be special-cased in the proposed scheme and we need an area "other".

Reply:
I have been trying to get this change merged for several years, so I am somewhat opposed to making more large architectural fixes. This PR is itself a rather cut-down version of the latest approach where I had added extra logic.

Currently I do not believe there is any "nice and simple" way to indicate the mapping between data type and object to use - for example

- source plot for DataPHA accepts lo/hi arguments unlike every-other plot [maybe we should add it to all plot types, but that's a separate discussion/issue/PR]
- plot_fit/plot_bkg_fit does not use the underlying "model" case, again only for DataPHA
- whatever is going on with template models

I've spent a lot of time trying to simplify this code and it has never ended in success: one conclusion is in fact automating the mapping between data-type and plot object logic doesn't really help.